pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -479,63 +394,101 @@ func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
 	return r.registryClient.Images().Get(dgst.String())
 }
 
-// getImageStreamTag retrieves the Image with tag `tag` for the ImageStream
-// associated with r.
-func (r *repository) getImageStreamTag(tag string) (*imageapi.ImageStreamTag, error) {
-	return r.registryClient.ImageStreamTags(r.namespace).Get(r.name, tag)
-}
-
 // getImageStreamImage retrieves the Image with digest `dgst` for the ImageStream
 // associated with r. This ensures the image belongs to the image stream.
 func (r *repository) getImageStreamImage(dgst digest.Digest) (*imageapi.ImageStreamImage, error) {
 	return r.registryClient.ImageStreamImages(r.namespace).Get(r.name, dgst.String())
 }
 
 // rememberLayers caches the provided layers
-func (r *repository) rememberLayers(manifest *schema1.SignedManifest, cacheName string) {
+func (r *repository) rememberLayers(manifest distribution.Manifest, cacheName string) {
 	if !r.pullthrough {
 		return
 	}
 	// remember the layers in the cache as an optimization to avoid searching all remote repositories
-	for _, layer := range manifest.FSLayers {
-		r.cachedLayers.RememberDigest(layer.BlobSum, cacheName)
+	for _, layer := range manifest.References() {
+		r.cachedLayers.RememberDigest(layer.Digest, cacheName)
 	}
 }
 
 // manifestFromImageWithCachedLayers loads the image and then caches any located layers
-func (r *repository) manifestFromImageWithCachedLayers(image *imageapi.Image, cacheName string) (*schema1.SignedManifest, error) {
-	manifest, err := r.manifestFromImage(image)
+func (r *repository) manifestFromImageWithCachedLayers(image *imageapi.Image, cacheName string, supportsSchema2 bool) (manifest distribution.Manifest, err error) {
+	// DockerConfigImage is not empty only in manifest 2.
+	isManifest2 := image.DockerConfigImage != ""
+
+	if supportsSchema2 {
+		if isManifest2 {
+			manifest, err = r.deserializedManifestFromImage(image)
+		} else {
+			manifest, err = r.signedManifestFromImage(image)
+		}
+	} else {
+		if isManifest2 {
+			err = fmt.Errorf("Unable to convert new image to old one")
```

Comment:
Quoting [upstream's compatibility](https://github.com/docker/distribution/blob/master/docs/compatibility.md#registry-v23) document:

> When the manifest is pulled by **tag** with Docker Engine 1.9 and older, the manifest is converted on-the-fly to Schema 1 and sent in the response. The Docker Engine 1.9 is compatible with this older format.

Conversion to older manifest schema is not supported only for the pull by **digest**. If this can be called only for *pull by digest*, it should be properly documented.